syft/artifact/id.go

```diff
@@ -0,0 +1,26 @@
+package artifact
+
+import (
+	"fmt"
+
+	"github.com/mitchellh/hashstructure/v2"
+)
+
+// ID represents a unique value for each package added to a package catalog.
+type ID string
+
+type Identifiable interface {
+	ID() ID
+}
+
+func DeriveID(obj interface{}) (ID, error) {
```

Comment:
Some of the larger comments on this PR talk about formalizing the ID across formats. I think a more strict signature might help with this. 

If we have this as `interface` what are all possible types within this project that could be passed here?
Can we limit it to some common field shared among all types?

Reply:
I though about doing something along those lines, where there would be a `IDFromLocation(...)` and `IDFromPackage(...)` in the `artifact` package. The first problem is that this would introduce a cyclic dependency between package and artifact --ideally it should be the artifact package that is at the "root". I then realized that all of the ID implementations would be the same anyway (pass an object into `Hash`). For this reason I went the direction of making a single utility function (also, I renamed this in a later commit to `IDFromHash(...)`).

The largest con against this decision I see is that it would be possible to get an `artifact.ID` for something that doesn't represent an artifact. However, since artifact.ID is a string, there is nothing stopping anyone from creating artifact.IDs --that is, there is no enforcement through encapsulation for this.

It could be that we move this function to a separate package... maybe `internal/hash`? Then at least there wouldn't be any problems with cyclic dependencies.